services/matching-service/start.ts

```diff
@@ -0,0 +1,120 @@
+import express from "express";
+import { Server } from "socket.io";
+import { createServer } from "http";
+
+import { extractCategory, extractDifficulty, isDuplicateID as checkIdExists } from "./domain/match_validators";
+
+const app = express();
+const httpServer = createServer(app);
+const io = new Server(httpServer);
+
+
+const port = process.env.PORT || 6001;
+
+io.on("connection", (socket) => {
+    socket.emit("reply", "hello from server");
+
+    socket.on("register", (msg) => {
+        console.log(`A NEW USER WITH ID ${msg.id} JOINED!`);    
+    
+        const difficulty = extractDifficulty(msg.difficulty);
+        const category = extractCategory(msg.category);
+        
+        // Check Valid Difficulty
+        if (difficulty == undefined) {
+            socket.emit("error", "Invalid Difficulty!");
+            return;
+        }
+        
+        //Check valid Category
+        if (category == undefined) {
+            socket.emit("error", "Invalid Category!");
+            return;
+        }
+        
+        // Check for duplicate ID and push to Queue
+        if (checkIdExists(msg.id)) {
+            socket.emit("error", "You are already in the queue!");
+            return;
+        } 
+
+        // Find match if exists, else create new queue and wait.
+        const this_match = {
+                id: msg.id,
+                difficulty: difficulty,
+                category: category,
+                sockAddr: socket.id
+            }
+        const queueName = difficulty + category; // e.g. easyrecursion
+        let this_queue = queues.get(queueName);
+        if (this_queue) {
+            // Match found, remove queue from queues
+            this_queue.push(this_match)
+            console.log("WE FOUND A MATCH!");
+            queues.delete(queueName);
```

Comment:
At any point there is only one user in the queue (hence why its now replaced with a single Match object instead of an array/queue). If a match is found, that queue is emptied. If the peer user is not connected, then the new user will take his place in the queue (delete and replace again).

Reply:
Hm but what if there are a few users that join the queue at once?